File: data_structure/stack/src/StackDemo.swift
// Swift
// データ構造: スタック (Stack)

class StackData {
    private var _data: [Int] = []
    
    func get() -> [Int] {
        return _data
    }
    
    func getIndex(item: Int) -> Int {
        if let index = _data.firstIndex(of: item) {
            return index
        } else {
            print("ERROR: \(item) は範囲外です")
            return -1
        }
    }
    
    func getValue(index: Int) -> Int? {
        if 0 <= index && index < _data.count {
            return _data[index]
        } else {
            print("ERROR: \(index) は範囲外です")
            return nil
        }
    }
    
    func push(item: Int) -> Bool {
        _data.append(item)
        return true
    }
    
    func pop() -> Bool {
        if !isEmpty() {
            _data.removeLast()
            return true
        } else {
            print("ERROR: 空です")
            return false
        }
    }
    
    func peek() -> Int? {
        return isEmpty() ? nil : _data.last
    }
    
    func isEmpty() -> Bool {
        return _data.isEmpty
    }
    
    func size() -> Int {
        return _data.count
    }
    
    func clear() -> Bool {
        _data.removeAll()
        return true
    }
}

func main() {
    print("Stack TEST -----> start")
    
    print("\nnew")
    let stackData = StackData()
    print("  現在のデータ: \(stackData.get())")
    
    print("\nis_empty")
    let output1 = stackData.isEmpty()
    print("  出力値: \(output1)")
    
    print("\nsize")
    let output2 = stackData.size()
    print("  出力値: \(output2)")
    
    print("\npush")
    let itemsToPush = [10, 20, 30, 40]
    for item in itemsToPush {
        print("  入力値: \(item)")
        let output = stackData.push(item: item)
        print("  出力値: \(output)")
        print("  現在のデータ: \(stackData.get())")
    }
    
    print("\nsize")
    let output3 = stackData.size()
    print("  出力値: \(output3)")
    
    print("\nis_empty")
    let output4 = stackData.isEmpty()
    print("  出力値: \(output4)")
    
    print("\npeek")
    let output5 = stackData.peek()
    print("  出力値: \(output5 ?? 0)")
    
    print("\nget_index")
    let input1 = 30
    print("  入力値: \(input1)")
    let output6 = stackData.getIndex(item: input1)
    print("  出力値: \(output6)")
    
    print("\nget_index")
    let input2 = 50
    print("  入力値: \(input2)")
    let output7 = stackData.getIndex(item: input2)
    print("  出力値: \(output7)")
    
    print("\npop")
    while !stackData.isEmpty() {
        let output = stackData.pop()
        print("  出力値: \(output)")
        print("  現在のデータ: \(stackData.get())")
    }
    
    print("\nis_empty")
    let output8 = stackData.isEmpty()
    print("  出力値: \(output8)")
    
    print("\nsize")
    let output9 = stackData.size()
    print("  出力値: \(output9)")
    
    print("\npop")
    let output10 = stackData.pop()
    print("  出力値: \(output10)")
    
    print("\npeek")
    let output11 = stackData.peek()
    print("  出力値: \(String(describing: output11))")
    
    print("\nStack TEST <----- end")
}

main()

File: data_structure/stack/src/StackDemo.go
// Go
// データ構造: スタック (Stack)

package main

import (
	"fmt"
)

type StackData struct {
	data []int
}

func (s *StackData) Get() []int {
	return s.data
}

func (s *StackData) GetIndex(item int) int {
	for i, v := range s.data {
		if v == item {
			return i
		}
	}
	fmt.Printf("ERROR: %d は範囲外です\n", item)
	return -1
}

func (s *StackData) GetValue(index int) *int {
	if 0 <= index && index < len(s.data) {
		return &s.data[index]
	}
	fmt.Printf("ERROR: %d は範囲外です\n", index)
	return nil
}

func (s *StackData) Push(item int) bool {
	s.data = append(s.data, item)
	return true
}

func (s *StackData) Pop() bool {
	if !s.IsEmpty() {
		s.data = s.data[:len(s.data)-1]
		return true
	}
	fmt.Println("ERROR: 空です")
	return false
}

func (s *StackData) Peek() *int {
	if !s.IsEmpty() {
		return &s.data[len(s.data)-1]
	}
	return nil
}

func (s *StackData) IsEmpty() bool {
	return len(s.data) == 0
}

func (s *StackData) Size() int {
	return len(s.data)
}

func (s *StackData) Clear() bool {
	s.data = []int{}
	return true
}

func main() {
	fmt.Println("Stack TEST -----> start")

	fmt.Println("\nnew")
	stackData := StackData{}
	fmt.Printf("  現在のデータ: %v\n", stackData.Get())

	fmt.Println("\nis_empty")
	output := stackData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsize")
	output2 := stackData.Size()
	fmt.Printf("  出力値: %v\n", output2)

	fmt.Println("\npush")
	itemsToPush := []int{10, 20, 30, 40}
	for _, item := range itemsToPush {
		fmt.Printf("  入力値: %v\n", item)
		output3 := stackData.Push(item)
		fmt.Printf("  出力値: %v\n", output3)
		fmt.Printf("  現在のデータ: %v\n", stackData.Get())
	}

	fmt.Println("\nsize")
	output4 := stackData.Size()
	fmt.Printf("  出力値: %v\n", output4)

	fmt.Println("\nis_empty")
	output5 := stackData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output5)

	fmt.Println("\npeek")
	output6 := stackData.Peek()
	if output6 != nil {
		fmt.Printf("  出力値: %v\n", *output6)
	} else {
		fmt.Printf("  出力値: %v\n", output6)
	}

	fmt.Println("\nget_index")
	input1 := 30
	fmt.Printf("  入力値: %v\n", input1)
	output7 := stackData.GetIndex(input1)
	fmt.Printf("  出力値: %v\n", output7)

	fmt.Println("\nget_index")
	input2 := 50
	fmt.Printf("  入力値: %v\n", input2)
	output8 := stackData.GetIndex(input2)
	fmt.Printf("  出力値: %v\n", output8)

	fmt.Println("\npop")
	for !stackData.IsEmpty() {
		output9 := stackData.Pop()
		fmt.Printf("  出力値: %v\n", output9)
		fmt.Printf("  現在のデータ: %v\n", stackData.Get())
	}

	fmt.Println("\nis_empty")
	output10 := stackData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output10)

	fmt.Println("\nsize")
	output11 := stackData.Size()
	fmt.Printf("  出力値: %v\n", output11)

	fmt.Println("\npop")
	output12 := stackData.Pop()
	fmt.Printf("  出力値: %v\n", output12)

	fmt.Println("\npeek")
	output13 := stackData.Peek()
	if output13 != nil {
		fmt.Printf("  出力値: %v\n", *output13)
	} else {
		fmt.Printf("  出力値: %v\n", output13)
	}

	fmt.Println("\nStack TEST <----- end")
}